linen_examples/cloud/README.md

```diff
@@ -0,0 +1,104 @@
+on Google Cloud, train an example on that VM and then shutting it down.
+
+The training is implemented in a shell that is run on the VM after startup by
+setting the `startup_script_file` in the metadata. The script opens a TMUX
+session, installs Flax repository from Github with all dependencies, and then
+runs the training in parallel with `gsutil rsync` that copies the training
+artifacts in a storage bucket.
+
+The advantage of this approach is that every training is run in a single VM
+that contains all code and configuration, so it is easy to run multiple
+experiments in parallel without interference. The individual trainings can also
+be inspected by logging into the VM via SSH and attaching to the tmux session.
+
+The script `launch_gce.py` launches the VM and prints out the relevant commands
+to track the progress update and to log into the machine.
+
+Note that the VM also shuts down if an error is encountered, after waiting for
+five minutes.
+
+## Preparation
+
+Prerequisites:
+
+1. Create a Google Cloud account.
+2. Set up billing: https://console.cloud.google.com/billing
+3. Create a storage bucket (GCS).
+4. Optional: Get quota for accelerators. This is usually granted with a short
+   delay: https://console.cloud.google.com/iam-admin/quotas
+
+## Setting up your environment
+
+The commands below use the same set of pre-defined environment variables.
+
+Mandatory environment variables:
+
+- `$PROJECT`: The name of your Google Cloud project.
+- `$GCS_BUCKET`: The name of the Google Cloud Storage bucket where the model
+  output (artifacts, final checkpoint) are to be stored.
+- `$ZONE`: Compute zone (e.g. `central1-a`).
+
+Optional environment variables:
+
+- `$REPO`: Alternative repo to use instead of the default
+  https://github.com/google/flax - this is useful for development.
+- `$BRANCH`: Alternative branch to use instead of the default `master`.
+
+## Training the MNIST example
+
+Use the following command to launch the MNIST example on cloud (make sure to set
+`$PROJECT` and `$GCS_BUCKET` accordingly):
+
+```shell
+python linen_examples/cloud/launch_gce.py \
+  --project=$PROJECT \
+  --zone=us-central1-a \
+  --machine_type=n2-standard-2 \
+  --gcs_workdir_base=gs://$GCS_BUCKET/workdir_base \
+  --repo=${REPO:-https://github.com/google/flax} \
+  --branch=${BRANCH:-master} \
+  --example=mnist \
+  --args='--config=configs/default.py' \
+  --name=default
+```
+
+## Training the imagenet example
+
+Note that you need to first prepare `imagenet2012` dataset. For this, download
+the data from http://image-net.org/ as described in the
+[tensorflow_datasets catalog](https://www.tensorflow.org/datasets/catalog/imagenet2012)
+and then prepare the dataset by running
+
+```shell
+python -c "
+import tensorflow_datasets as tfds
+tfds.builder('imagenet2012').download_and_prepare(
+    download_config=tfds.download.DownloadConfig(
+        manual_dir='$IMAGENET_DOWNLOAD_PATH'))
```

Comment:
Is the user supposed to set $IMAGENET_DOWNLOAD_PATH? That's not specified anywhere, it seems?

Reply:
Or is it actually `.` (current directory)?

Reply:
Explained this better.